chronos/achannels.nim

```diff
@@ -0,0 +1,254 @@
+#
+#    Chronos multi-threading asynchronous channels
+#             (c) Copyright 2018-Present
+#         Status Research & Development GmbH
+#
+#              Licensed under either of
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+import strutils
+import asyncloop, handles, transport, asyncsync
+
+const hasThreadSupport* = compileOption("threads")
+
+when hasThreadSupport:
+  import locks
+
+type
+  RawAsyncChannelImpl {.pure, final.} = object
+    rd, wr, count, mask: int
+    maxItems: int
+    refCount: int
+    data: ptr UncheckedArray[byte]
+    when hasThreadSupport:
+      lock: Lock
+    eventNotEmpty: AsyncThreadEvent
+    eventNotFull: AsyncThreadEvent
+
+  RawAsyncChannel = ptr RawAsyncChannelImpl
+  AsyncChannel*[Msg] = RawAsyncChannel
+  AsyncChannelError* = object of CatchableError
+
+proc initLocks(rchan: RawAsyncChannel) {.inline.} =
+  ## Initialize and create OS locks.
+  when hasThreadSupport:
+    initLock(rchan.lock)
+  else:
+    discard
+  rchan.eventNotEmpty = newAsyncThreadEvent()
+  if rchan.maxItems > 0:
+    rchan.eventNotFull = newAsyncThreadEvent()
+
+proc deinitLocks(rchan: RawAsyncChannel) {.inline.} =
+  ## Deinitialize and close OS locks.
+  when hasThreadSupport:
+    deinitLock(rchan.lock)
+  else:
+    discard
+
+  close(rchan.eventNotEmpty)
+  if rchan.maxItems > 0:
+    close(rchan.eventNotFull)
+
+proc acquireLock(rchan: RawAsyncChannel) {.inline.} =
+  ## Acquire lock in multi-threaded mode and do nothing in
+  ## single-threaded mode.
+  when hasThreadSupport:
+    acquire(rchan.lock)
+  else:
+    discard
+
+proc releaseLock(rchan: RawAsyncChannel) {.inline.} =
+  ## Release lock in multi-threaded mode and do nothing in
+  ## single-threaded mode.
+  when hasThreadSupport:
+    release(rchan.lock)
+  else:
+    discard
+
+proc newAsyncChannel*[Msg](maxItems: int = -1): AsyncChannel[Msg] =
+  ## Create new AsyncChannel[Msg] with internal queue size ``maxItems``.
+  ##
+  ## If ``maxItems <= 0`` (default value), then queue size is unlimited.
+  let loop = getGlobalDispatcher()
+  result = cast[AsyncChannel[Msg]](allocShared(sizeof(RawAsyncChannelImpl)))
```

Comment:
Is this done to avoid GC-safety warnings? A regular heap allocation with `new()` would allow setting a finaliser to close the channel and it's very unlikely that it would be garbage collected before we're done with it.

Reply:
Nope allocation in shared heap was made to allow all the threads access this memory without any problems. If you allocate heap with `new` you will allocate in thread's local heap.

Reply:
Yes, but as long as that thread-local heap is not collected and access to it is guarded by locks, it's GC safe, even if the compiler can't verify it.